### DISCUSSION TITLE
Train Custom Data - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#2795] [Docs]
# Train Custom Data - Ultralytics YOLOv8 Docs

Train your custom dataset with YOLOv5. Learn to collect, label and annotate images, and train and deploy models. Get started now.

https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/

**Comment by 19T1021243** [comment]
Please let me know how to make the model recognize that my label is a *.json file when training the dataset

**Reply by glenn-jocher** [reply]
If your labels are in COCO JSON format you must export them first to YOLO format. You can use https://github.com/ultralytics/JSON2YOLO

**Comment by ocomputer** [comment]
Hi! I have a project where I need to 1) change the background on yolov5 to just one solid color as well as 2) blur the background to help just identify certain objects. I already trained my data set to identify the objects that I want, but i don't know how to complete the other tasks. please help!

**Reply by glenn-jocher** [reply]
@ocomputer the default background color for YOLOv5 is a grey color defined as the Imagenet mean color. You can modify this default if you'd like the YOLOv5 dataloader in https://github.com/ultralytics/yolov5/blob/master/utils/dataloaders.py.

**Comment by aish0512** [comment]
Hey, how to get the bounding boxes of the object using yolov5?

**Reply by glenn-jocher** [reply]
The boxes are in the results object returned by the model, i.e.:

```python
results = model('path/to/image.jpg')
```

For details see https://docs.ultralytics.com/modes/predict/

**Comment by Codilex** [comment]
can anyone please tell me that how i can train 50k dataset in splits, like want to train first 5k data then next 5k and so on , so that it can consume less time at moment. I want to create only one detection model which keep updating when i add new dataset. it should not take double time if add new set of data for example for first 5k data it took 1 hr so it should not take 2 hr for adding next 5k data it should start updating the model from new set and should take around 1 hr

**Reply by pderrenger** [reply]
@Codilex to train your dataset in increments and update your model with new data without retraining from scratch each time, you can use the following approach:

1. **Initial Training**: Start by training your YOLOv8 model on the first 5k dataset. Save the weights after training.

    ```python
    from ultralytics import YOLO

    # Load a pretrained YOLO model
    model = YOLO('yolov8n.pt')

    # Train the model on the first 5k dataset
    results = model.train(data='first_5k_data.yaml', epochs=3)
    ```

2. **Incremental Training**: For each subsequent 5k dataset, load the weights from the previous training session and continue training. This is known as fine-tuning or transfer learning.

    ```python
    # Load the model with the weights from the previous training
    model = YOLO('path/to/last_weights.pt')

    # Train the model on the next 5k dataset
    results = model.train(data='next_5k_data.yaml', epochs=3)
    ```

3. **Resume Training**: If you want to resume training from the last checkpoint, you can use the `resume` flag.

    ```python
    # Resume training from the last checkpoint
    model = YOLO('path/to/last_checkpoint.pt')
    results = model.train(resume=True)
    ```

By using this method, you avoid retraining from scratch each time you add new data. The model will update with the new data, and the training time for each new dataset should be similar to the initial training time, assuming the dataset size and complexity are consistent.

Remember to adjust the `data` parameter to point to the correct dataset configuration file for each training session. Also, ensure that your datasets are prepared correctly and that the configuration files reflect the new data splits.

Keep in mind that as your model sees more data, the training time might increase slightly due to the model converging and the optimizer updates becoming more nuanced. However, it should not double as you described.

**Comment by enekuie** [comment]
is there any change in new ultralytics update about custom dataset directory? i updated it recently but it says can't find my dataset yaml file when it was working properly in pervious versions of ultralytics

**Reply by pderrenger** [reply]
@enekuie it sounds like you're encountering an issue with locating your dataset YAML file after updating. There have been no intentional changes to the way custom dataset directories are handled in recent updates that should cause this issue. However, it's important to ensure that the path to your dataset YAML file is correct and that the file is indeed in the expected directory.

Here are a few steps you can take to troubleshoot the issue:

1. **Verify the YAML file path**: Ensure that the path to your dataset YAML file is correct. If you're using a relative path, it should be relative to the directory from which you're running the training command.

2. **Check for typos**: Double-check the file name and path for any typos or case sensitivity issues.

3. **File permissions**: Make sure that the YAML file has the appropriate read permissions.

4. **YAML file structure**: Confirm that the YAML file is properly formatted and includes the necessary fields such as `train`, `val`, `nc`, and `names`.

5. **Working directory**: Ensure that your current working directory is as expected. Sometimes, running scripts from different directories can cause path issues.

6. **Version-specific changes**: While there shouldn't be any breaking changes, it's always a good idea to review the release notes for the version you've updated to, in case there are any relevant notes about changes to dataset handling.

If you continue to face issues, you might want to check the documentation for any updates on custom dataset training or post an issue on the GitHub repository with details of the error message and your directory structure for further assistance. 🛠️

Remember to always keep a backup of your working environment and dataset configurations to revert back to a previous state if necessary. 🔄

**Comment by FengRongYue** [comment]
Hi!How can i change sapcial layout of anchors?

**Reply by pderrenger** [reply]
@FengRongYue to adjust the spatial layout of anchors in YOLOv8, you can modify the anchor shapes directly in your model's YAML configuration file. Here's a concise guide on how to do it:

1. **Analyze Your Dataset**: Use the `analyze` function to compute optimal anchors for your dataset. This can be done after you've accumulated your training images and annotations.

2. **Modify YAML File**: Open the YAML file that defines your model architecture. Look for the `anchors` key in the file.

3. **Edit Anchors**: Replace the existing anchor values with the ones suggested by the `analyze` function. Anchors are typically defined in pairs of width and height values.

4. **Train Your Model**: With the updated anchors, you can now train your model using the `train` mode with your custom dataset.

Remember, the anchors should reflect the scale and aspect ratio of the objects in your dataset for optimal performance. If you need further guidance, please refer to the YOLOv8 documentation on model configuration and training. 📘🚀

**Comment by kaanicer01** [comment]
Hi,
I think the yolov5s.pt model is trained for coco dataset(80 classes). But I want to train VOC dataset which have 20 classes. Is there any other weights for I can use for VOC and other datasets, or is yolov5s is applicable for python train.py "--data data/VOC.yaml --weights yolov5s.pt" command? Or for instance, imageNet have classes more than 80, how can I use yolov5s.pt for training Imagenet?

**Reply by pderrenger** [reply]
@kaanicer01 hello! It seems there's a bit of confusion in your query regarding YOLOv5 and YOLOv8. Since you're asking about YOLOv8, I'll focus on that.

For training a YOLOv8 model on a dataset like VOC, which has a different number of classes than COCO, you can certainly start with a pretrained model such as `yolov8n.pt`. The pretrained weights provide a good starting point even if the number of classes differs. You'll need to modify your dataset YAML file to reflect the correct number of classes and provide the paths to your VOC dataset.

Here's how you can train a YOLOv8 model on the VOC dataset:

1. Prepare your VOC dataset in the correct format.
2. Create a `VOC.yaml` file that specifies the paths to your train/val data, the number of classes (20 for VOC), and class names.
3. Use the following command to start training:

```python
from ultralytics import YOLO

# Load a pretrained YOLOv8 model
model = YOLO('yolov8n.pt')

# Train the model on your VOC dataset
results = model.train(data='VOC.yaml', epochs=50)
```

For datasets with a different number of classes than the pretrained model (like ImageNet), the same approach applies. You'll need to adjust your dataset YAML file accordingly. The pretrained weights will still be beneficial as they have learned a variety of features from the COCO dataset that can be transferable to other datasets.

Remember to check the [Ultralytics Docs](https://docs.ultralytics.com) for more detailed instructions on training with custom datasets. 😊🚀

**Comment by mtemgoua** [comment]
Hello
I am a student working on a project: obstacles in maritime settings 

where I want to use the following challenge: https://macvi.org/workshop/macvi24/challenges/usv_det

to develop a model for object detection, specifically with obstacles in maritime settings. For a baseline, I want to train YOLO (https://pytorch.org/hub/ultralytics_yolov5/) on the challenge dataset and make some improvement. 

I was wondering if you could please help me.
Thank you for your consideration and for your time doing this.
Warm regards,

**Reply by pderrenger** [reply]
@mtemgoua hello! It's great to hear about your project on obstacle detection in maritime settings. For training a model on your specific dataset, you'll want to use YOLOv8, which is the latest version in the YOLO series and offers several improvements over previous versions.

To get started with training YOLOv8 on your custom dataset, you'll need to follow these general steps:

1. **Collect and Prepare Your Dataset**: Make sure your dataset is labeled correctly. You'll need images and corresponding annotations in a format that YOLO can understand, typically in a `.txt` file for each image with bounding box coordinates and class labels.

2. **Set Up Your Environment**: Install the necessary dependencies and set up your Python environment to work with YOLOv8. You can find the installation instructions in the Ultralytics Docs.

3. **Configure Your Training**: Create a `.yaml` file for your dataset, specifying the paths to your training and validation data, the number of classes, and class names.

4. **Train Your Model**: Use the YOLOv8 Python interface to train your model on your custom dataset. You can start with a pretrained model to speed up the training process and potentially improve your results.

5. **Evaluate Your Model**: After training, evaluate your model's performance on a validation set to see how well it's doing.

6. **Inference**: Once you're satisfied with the training, you can use your trained model to perform object detection on new images or videos.

Here's a simplified example of how you might train a YOLOv8 model on your custom dataset:

```python
from ultralytics import YOLO

# Load a pretrained YOLOv8 model
model = YOLO('yolov8n.pt')

# Train the model on your custom dataset
results = model.train(data='your_dataset.yaml', epochs=50)
```

For detailed instructions and examples, please refer to the [Train](https://docs.ultralytics.com/modes/train) section of the Ultralytics Docs. There you'll find comprehensive guidance on preparing your data, configuring your training, and running the training process.

If you encounter any issues or have further questions, feel free to reach out. Good luck with your project! 🚢🔍

**Comment by samuelsaqueiroz** [comment]
Why do you guys use the some folder to train/validation? The same images to both train and validate? That's biased.

**Reply by pderrenger** [reply]
@samuelsaqueiroz hello! In our documentation and examples, we often use a small dataset like COCO128 for simplicity and demonstration purposes. This dataset is a subset of the larger COCO dataset and is designed to be a quick and easy way to get started with training and validating models.

However, for a proper training setup, it's crucial to have separate datasets for training and validation. This ensures that the model is evaluated on images it has never seen during training, which provides a more accurate measure of its performance and generalization capabilities.

When you're training your own model with YOLOv8, you should definitely use distinct sets of images for training and validation. The data configuration file (`data.yaml`) allows you to specify separate paths for the training and validation datasets to facilitate this.

If you have any further questions or need assistance with setting up your custom training and validation datasets, feel free to reach out! 😊👍

**Comment by naveenvj25** [comment]
Is it possible to directly give labels in JSON format to train the Yolo v5 model? My labels are in labelme json format and I am not able to properly convert to Yolo text file.

**Reply by pderrenger** [reply]
@naveenvj25 for YOLOv8, the training process typically requires labels in a specific text format, where each line corresponds to an object and contains class index and bounding box coordinates in normalized xywh format (center x, center y, width, height). However, if you have labels in JSON format from labelme, you would need to convert them to the expected YOLO format before training.

While direct JSON input is not supported, you can write a custom script to convert your labelme JSON annotations to the YOLO text format. The conversion process involves reading the JSON file, extracting the bounding box information, and transforming it into the YOLO format.

Here's a brief example of how you might convert labelme JSON to YOLO format:

```python
import json
import os

def convert_labelme_json_to_yolo(json_path, output_dir, image_width, image_height):
    with open(json_path) as f:
        data = json.load(f)
    
    yolo_labels = []
    for shape in data['shapes']:
        if shape['shape_type'] != 'rectangle':
            print(f"Skipping non-rectangle shape type: {shape['shape_type']}")
            continue
        
        class_name = shape['label']
        # Add your class to index mapping here
        class_index = class_name_to_index(class_name)
        
        points = shape['points']
        # Calculate YOLO format coordinates
        x_center = (points[0][0] + points[1][0]) / 2 / image_width
        y_center = (points[0][1] + points[1][1]) / 2 / image_height
        width = abs(points[1][0] - points[0][0]) / image_width
        height = abs(points[1][1] - points[0][1]) / image_height
        
        yolo_labels.append(f"{class_index} {x_center} {y_center} {width} {height}")
    
    # Save YOLO labels to a .txt file
    base_name = os.path.splitext(os.path.basename(json_path))[0]
    output_path = os.path.join(output_dir, f"{base_name}.txt")
    with open(output_path, 'w') as f:
        f.write("\n".join(yolo_labels))

def class_name_to_index(class_name):
    # Define your mapping from class name to YOLO index here
    class_mapping = {'class1': 0, 'class2': 1} # Example mapping
    return class_mapping.get(class_name, -1) # Returns -1 if class_name is not found

# Example usage
convert_labelme_json_to_yolo('path/to/your/label.json', 'path/to/output/dir', 1920, 1080)
```

Please adjust the `class_name_to_index` function to map your specific class names to their corresponding indices as per your dataset classes.

For more details on training with custom data in YOLOv8, you can refer to the [Train Custom Data](https://docs.ultralytics.com/modes/train) section of the Ultralytics Docs. Remember to adapt the instructions to YOLOv8 specifics. 😊🚀

**Reply by naveenvj25** [reply]
Thank you for the information ,  I will check this

**Reply by glenn-jocher** [reply]
@naveenvj25 you're welcome! If you have any more questions or need further assistance with YOLOv8 or the conversion process, feel free to reach out. Good luck with your model training! 😊👍

**Comment by rampravesh80** [comment]
Hello,
I've successfully trained a YOLOv5 model using Google Colab, but I'm encountering issues when attempting the same process in a Jupyter Notebook. My preference for Jupyter Notebook stems from the limitation of Google Colab sessions, which expire and result in data loss. My project involves training YOLOv5 on a large dataset, expected to exceed 12 hours, a duration not feasible on Colab due to the mentioned session timeouts. Below is an error I encountered while using Jupyter Notebook. Could you please provide guidance on resolving this?
Error:
train: weights=yolov5s.pt, cfg=, data=C:/Users/rampr/Documents/yolov5/data/coco128.yaml, hyp=yolov5\data\hyps\hyp.scratch-low.yaml, epochs=100, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, evolve_population=yolov5\data\hyps, resume_evolve=None, bucket=, cache=None, image_weights=False, device=, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=yolov5\runs\train, name=exp, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, seed=0, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest, ndjson_console=False, ndjson_file=False
github: up to date with https://github.com/ultralytics/yolov5 
YOLOv5  v7.0-277-g703d37e Python-3.9.7 torch-2.1.2+cu118 CUDA:0 (NVIDIA GeForce GTX 1650 Ti, 4096MiB)

hyperparameters: lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
TensorBoard: Start with 'tensorboard --logdir yolov5\runs\train', view at http://localhost:6006/
COMET WARNING: Comet credentials have not been set. Comet will default to offline logging. Please set your credentials to enable online logging.
COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: tensorflow, torch, tensorboard, keras.
COMET INFO: Using 'C:\\Users\\rampr\\Documents\\.cometml-runs' path as offline directory. Pass 'offline_directory' parameter into constructor or set the 'COMET_OFFLINE_DIRECTORY' environment variable to manually choose where to store offline experiment archives.
COMET WARNING: Native output logging mode is not available, falling back to basic output logging
Traceback (most recent call last):
  File "C:\Users\rampr\Documents\yolov5\train.py", line 836, in <module>
    main(opt)
  File "C:\Users\rampr\Documents\yolov5\train.py", line 616, in main
    train(opt.hyp, opt, device, callbacks)
  File "C:\Users\rampr\Documents\yolov5\train.py", line 147, in train
    loggers = Loggers(
  File "C:\Users\rampr\Documents\yolov5\utils\loggers\__init__.py", line 146, in __init__
    self.comet_logger = CometLogger(self.opt, self.hyp)
  File "C:\Users\rampr\Documents\yolov5\utils\loggers\comet\__init__.py", line 97, in __init__
    self.data_dict = self.check_dataset(self.opt.data)
  File "C:\Users\rampr\Documents\yolov5\utils\loggers\comet\__init__.py", line 239, in check_dataset
    data_config = yaml.safe_load(f)
  File "C:\Users\rampr\anaconda3\lib\site-packages\yaml\__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
  File "C:\Users\rampr\anaconda3\lib\site-packages\yaml\__init__.py", line 79, in load
    loader = Loader(stream)
  File "C:\Users\rampr\anaconda3\lib\site-packages\yaml\loader.py", line 34, in __init__
    Reader.__init__(self, stream)
  File "C:\Users\rampr\anaconda3\lib\site-packages\yaml\reader.py", line 85, in __init__
    self.determine_encoding()
  File "C:\Users\rampr\anaconda3\lib\site-packages\yaml\reader.py", line 124, in determine_encoding
    self.update_raw()
  File "C:\Users\rampr\anaconda3\lib\site-packages\yaml\reader.py", line 178, in update_raw
    data = self.stream.read(size)
  File "C:\Users\rampr\anaconda3\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 297: character maps to <undefined>
COMET INFO: Couldn't find a Git repository in 'C:\\Users\\rampr\\Documents' nor in any parent directory. Set `COMET_GIT_DIRECTORY` if your Git Repository is elsewhere.
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml OfflineExperiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     url                   : [OfflineExperiment will get URL after upload]
COMET INFO:   Others:
COMET INFO:     Name               : exp
COMET INFO:     offline_experiment : True
COMET INFO:   Uploads:
COMET INFO:     conda-environment-definition : 1
COMET INFO:     conda-info                   : 1
COMET INFO:     conda-specification          : 1
COMET INFO:     environment details          : 1
COMET INFO:     installed packages           : 1
COMET INFO: 
COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: tensorflow, torch, tensorboard, keras.
COMET INFO: Still saving offline stats to messages file before program termination (may take up to 120 seconds)
COMET INFO: Starting saving the offline archive
COMET INFO: To upload this offline experiment, run:
    comet upload C:\Users\rampr\Documents\.cometml-runs\0e3c762fa7864714b3ee53d3f22f6091.zip

**Reply by pderrenger** [reply]
@rampravesh80 it seems there's a bit of confusion in your message. You're referring to YOLOv5, but this is the YOLOv8 repository. However, I can still provide some general advice that might help you resolve the issue you're encountering in Jupyter Notebook.

The error you're seeing, `UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 297: character maps to <undefined>`, suggests that there's a problem with character encoding when reading a file. This often happens when a file is not encoded in UTF-8 or there's an unexpected character that can't be mapped correctly.

Here are a few steps you can take to troubleshoot this issue:

1. **Check File Encoding**: Ensure that your YAML file (`coco128.yaml`) is saved with UTF-8 encoding. You can usually set the encoding in your text editor or IDE.

2. **Specify Encoding in Python**: When opening files in Python, you can specify the encoding to use. For example:
   ```python
   with open('yourfile.yaml', 'r', encoding='utf-8') as file:
       data = yaml.safe_load(file)
   ```

3. **Update Your Code**: If you're using code from the YOLOv5 repository, make sure it's up to date and compatible with the version of YOLO you're using.

4. **Environment Consistency**: Ensure that your local environment matches the one you used in Google Colab as closely as possible, including Python version, library versions, etc.

5. **Comet.ml Logging**: The warnings from Comet.ml suggest that it's set to offline mode. If you're not using Comet.ml for experiment tracking, you can remove or comment out the related code.

6. **File Paths**: Double-check the file paths in your configuration file (`coco128.yaml`). Windows uses backslashes `\` for paths, which can sometimes cause issues. You might need to use raw strings (prefix the string with `r`) or replace backslashes with forward slashes `/`.

7. **Session Persistence**: For long training sessions, consider setting up your own environment with session persistence, such as using a local machine, a dedicated server, or a cloud instance with a longer timeout or no timeout at all.

If you continue to have issues specific to YOLOv8, please provide the exact error message related to YOLOv8, and I'll be happy to help you further. Remember to always use the latest documentation and resources for YOLOv8 to ensure compatibility and up-to-date instructions. 📘🛠️
